pkg/controllers/node/ipam.go

```diff
@@ -501,7 +521,62 @@ func (c *ipamController) updateMetrics() {
 	log.Debug("IPAM metrics updated")
 }
 
-// reviewIPAMState scans Calico IPAM and determines if any IPs appear to be leaks, and if any nodes should have their
+// checkEmptyBlocks looks at known empty blocks, and releases their affinity
+// if appropriate. A block is a candidate for having its affinity released if:
+// - The block is empty.
+// - The block's node has at least one other affine block.
+// - The other blocks on the node are not at capacity.
+func (c *ipamController) checkEmptyBlocks() error {
+	for blockCIDR, node := range c.emptyBlocks {
+		logc := log.WithFields(log.Fields{"blockCIDR": blockCIDR, "node": node})
+		nodeBlocks := c.blocksByNode[node]
+		if len(nodeBlocks) > 1 {
+			// The node has more than one block. Check that the other blocks allocated to this node
+			// are not at capacity. We only release blocks when there's room in the other affine blocks,
+			// otherwise the next IP allcation will just assign a new block to this node anyway.
+			numAddressesAvailableOnNode := 0
+			for b := range nodeBlocks {
+				if b == blockCIDR {
+					// Skip the known empty block.
+					continue
+				}
+
+				// Sum the number of unallocated addresses across the other blocks.
+				kvp := c.allBlocks[b]
+				numAddressesAvailableOnNode += len(kvp.Value.(*model.AllocationBlock).Unallocated)
+			}
+
+			// Make sure there are some addresses available before releasing.
+			if numAddressesAvailableOnNode < 3 {
+				logc.Debug("Block is still needed, skip release")
+				continue
+			}
+
+			// We can release the empty one.
+			logc.Infof("Releasing affinity for empty block (node has %d total blocks)", len(nodeBlocks))
+			_, cidr, err := cnet.ParseCIDR(blockCIDR)
+			if err != nil {
+				return err
+			}
+			err = c.client.IPAM().ReleaseAffinity(context.TODO(), *cidr, node, true)
+			if err != nil {
+				return err
```

Comment:
Looks like `ReleaseAffinity` can return `errBlockNotEmpty` if the block is no longer empty.  I think that should be handled with a continue to look at the next host.

In general, you might want to continue on error in this loop so that each block has a chance at getting freed.